File: Client/src/app/model/person.ts
/**
 * Keep track of a single person in the Database.
 */

import { ImageManager } from "../ServerInterfaces/ImageManager";

export class Person {
    
  /**
   * Unique person ID field assigned by the database
   */
    personId: string;

    /**
     *  Person's address as HTML */
    addressAsHTML: string;

    /** Address as a multi-line text field */
    addressAsText: string;

    /** Address as a comma seperated text field */
    addressCommaSeparated: string;

    /** Person's first name (required) */
    firstName: string;

    /** Person's last name (required) */
    lastName: string;

    /** First line of Person's Address */
    addressLine1: string;

    /** Second line of Person's Address */
    addressLine2: string;

    /** City Person resides in */
    city: string;

    /** State or Territory Person resides in */
    stateOrTerritory: string;

    /** Country Peson resides in */
    country: string;

    /** Person's Zip code */
    zipCode: string;

    /** Person's Birthdate (as Date) */
    birthDate: Date;

    /** Person's Birthdate (as String) */
    displayBirthDate: string;

    /** Calculated age of person based on date of birth and todays date */
    age: string;

    /** Database GUID of the image to display for the person.  "0" means no Image */
    imageGUID: string;

    /** Is the person's entry on the UI highlighted */
    highlight: boolean;

    /** Are we showing the details for the person on the UI */
    showDetails: boolean;

    /** Are we editing the person on the UI */
    editDetails: boolean; 

    /** Set of interests (Array<string>) that the person has */
    interests: Array<string>;

    /** Create a new person with no interests */
    constructor (){
      this.interests = new Array<string>();
    }

    /** Returns the image of the Person to use on the UI */
    Photo ()
    {
      return ImageManager.GetImageUrl ( this.imageGUID ) ;
    }

    /**
     * Copy the values from one person to another
     * @param source Source of the values to copy to 'this'
     */
    CopyValues ( source:Person )
    {
        this.firstName = source.firstName ;
        this.lastName = source.lastName;
        this.addressLine1 = source.addressLine1 ;
        this.addressLine2 = source.addressLine2 ;
        this.addressAsHTML = source.addressAsHTML ;
        this.addressAsText = source.addressAsText;
        this.addressCommaSeparated = source.addressCommaSeparated ;
        this.city = source.city ;
        this.stateOrTerritory = source.stateOrTerritory ;
        this.country = source.country ;
        this.zipCode = source.zipCode ;
        this.birthDate = source.birthDate ;
        this.displayBirthDate = source.displayBirthDate ;
        this.age = source.age ;
      }
}

File: PeopleSearchServer/PeopleSearchServer/Models/Person.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;

namespace PeopleSearchServer.Models
{
    /// <summary>
    /// Model for the Person Database Entry.  This describes a single person.
    /// </summary>
    public class Person
    {
        /// <summary>
        /// Unique identifier (Key) for database
        /// </summary>
        [Key]
        public int PersonId
        {
            get;
            set;
        }

        /// <summary>
        /// Person's First Name
        /// </summary>
        [Required]
        [Display(Name = "First Name")]
        public string FirstName
        {
            get;
            set;
        }

        /// <summary>
        /// Person's Last Name
        /// </summary>
        [Required]
        [Display(Name = "Last Name")]
        public string LastName
        {
            get;
            set;
        }

        /// <summary>
        /// Date of Birth
        /// </summary>
        [DataType(DataType.Date)]
        [Display(Name = "Birth Date")]
        public DateTime BirthDate
        {
            get;
            set;
        }

        [NotMapped]
        public int Age
        {
            get
            {
                TimeSpan howLongAlive = DateTime.Now - BirthDate ;

                return (int) Math.Floor ( howLongAlive.TotalDays / 365.0);
            }
        }

        [NotMapped]
        public string DisplayBirthDate
        {
            get
            {
                return BirthDate.ToString("MM/dd/yyyy");
            }
        }

        /// <summary>
        /// GUID of Image of person
        /// </summary>
        [Display(Name = "Image GUID")]
        public string ImageGUID
        {
            get;
            set;
        }

        /// <summary>
        /// Set of interests for the person
        /// </summary>
        public string[] Interests
        {
            get
            {
                List<string> resultsList = new List<string>();

                using (var context = new PeopleSearchContext())
                {
                    var interestsQuery = context.Interest.Where(interest => interest.Person.PersonId == this.PersonId);

                    foreach (Interest interest in interestsQuery)
                    {
                        resultsList.Add(interest.TheInterest);
                    }
                }

                return resultsList.ToArray();
            }
        }

        #region Address

        /// <summary>
        /// Address Line 1
        /// </summary>
        [Display(Name = "Address Line 1")]
        public string AddressLine1
        {
            get;
            set;
        }

        /// <summary>
        /// Address Line 2
        /// </summary>
        [Display(Name = "Address Line 2")]
        public string AddressLine2
        {
            get;
            set;
        }

        /// <summary>
        /// City of Residence
        /// </summary>
        public string City
        {
            get;
            set;
        }

        /// <summary>
        /// State or Territory of Residence
        /// </summary>
        [Display(Name = "State or Territory")]
        public string StateOrTerritory
        {
            get;
            set;
        }

        [Display(Name = "Zip Code")]
        public string ZipCode
        {
            get;
            set;
        }

        /// <summary>
        /// Country of Residence
        /// </summary>
        public string Country
        {
            get;
            set;
        }

        #endregion Address

        #region Address Formatting

        /// <summary>
        /// Format the address using the parts seperator passed in ('\n', '<br/>', ','  ...)
        /// </summary>
        /// <param name="lf">Line Separator</param>
        /// <returns>Formatted address</returns>
        private string FormatAddress(string lf)
        {
            StringBuilder results = new StringBuilder();

            string cityState = FormatAddressPart(City, (string.IsNullOrEmpty(StateOrTerritory) ? "" : ", ")) +
                FormatAddressPart(StateOrTerritory, "");

            results.Append(FormatAddressPart(AddressLine1, lf));
            results.Append(FormatAddressPart(AddressLine2, lf));
            results.Append(FormatAddressPart(cityState, lf));
            results.Append(FormatAddressPart(ZipCode, lf));
            results.Append(Country);

            return results.ToString();
        }

        /// <summary>
        /// IF there is text in the address part, then return the text followed by the separator.
        /// Otherwise return the empty string.
        /// </summary>
        /// <param name="part">Part of the address to go on one line</param>
        /// <param name="linefeed">Line parts separator</param>
        /// <returns>Formatted address with separator in it.</returns>
        private string FormatAddressPart(string part, string linefeed)
        {
            if (string.IsNullOrWhiteSpace(part))
            {
                return "";
            }
            else
            {
                return $"{part}{linefeed}";
            }
        }

        #endregion


        /// <summary>
        /// Address of the preson as Line Separated string
        /// </summary>
        [Display(Name = "Address")]
        [NotMapped]
        public string AddressAsText
        {
            get
            {
                return FormatAddress( Environment.NewLine );
            }
        }
        /// <summary>
        /// Address of the preson as HTML
        /// </summary>
        [Display(Name = "Address")]
        [NotMapped]
        public string AddressAsHTML
        {
            get
            {
                return FormatAddress( "<br/>" );
            }
        }

        /// <summary>
        /// Address of the preson as HTML
        /// </summary>
        [Display(Name = "Address")]
        [NotMapped]
        public string AddressCommaSeparated
        {
            get
            {
                return FormatAddress(", ");
            }
        }

        /// <summary>
        /// Constructor... Create an empty person
        /// </summary>
        public Person ()
        {
            this.FirstName = string.Empty;
            this.LastName = string.Empty;
            this.AddressLine1 = string.Empty;
            this.AddressLine2 = string.Empty;
            this.BirthDate = new DateTime ( 2000, 1, 1 );
            this.City = string.Empty;
            this.Country = string.Empty;
            this.FirstName = string.Empty;
            this.LastName = string.Empty;
            this.StateOrTerritory = string.Empty;
            this.ZipCode = string.Empty;
        }

        /// <summary>
        /// String representation of person for debugging
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return $"Person: {FirstName} {LastName} : DOB: {DisplayBirthDate}";
        }
    }
}
